src/back.jl

```diff
@@ -14,51 +14,65 @@ init_grad(x) = zero(x)
 zero_grad!(x) = zero(x)
 zero_grad!(x::AbstractArray) = (x .= 0)
 
-scan(c::Call) = foreach(scan, c.args)
-
-function scan(x::Tracked)
-  x.isleaf && return
-  ref = x.ref += 1
-  if ref == 1
-    scan(x.f)
-    isdefined(x, :grad) && (x.grad = zero_grad!(x.grad))
+function _walk(queue, seen, c::Call)
+  foreach(c.args) do x
+    x === nothing && return
+    id = objectid(x)
+    if id ∉ seen
+      push!(seen, id)
+      pushfirst!(queue, x)
+    end
+    return
   end
-  return
 end
 
-function scan(x)
-  istracked(x) && scan(tracker(x))
-  return
+function walk(f, x::Tracked, seen = Set{UInt64}(); once = true)
+  queue = Tracked[x]
+  while !isempty(queue)
+    x = pop!(queue)
+    f(x, seen)
+    _walk(queue, seen, x.f)
+    once && !x.isleaf && (x.f = Call(missing, ()))
+  end
 end
-
-function back_(c::Call, Δ, once)
+  
+function back_(c::Call, Δ, seen)
   Δs = c.func(Δ)
   (Δs isa Tuple && length(Δs) >= length(c.args)) ||
     error("Gradient is not a tuple of length $(length(c.args))")
-  foreach((x, d) -> back(x, d, once), c.args, data.(Δs))
+  foreach(c.args) do x
+    isdefined(x, :grad) || return
+    objectid(x) ∉ seen && zero_grad!(x.grad)
+  end
+  foreach((x, d) -> back_(x, d), c.args, data.(Δs))
 end
 
-back_(::Call{Nothing}, Δ, once) = nothing
-back_(::Call{Missing}, Δ, once) = error("`back!` was already used")
+back_(::Call{Nothing}, Δ, seen) = nothing
+back_(::Call{Missing}, Δ, seen) = error("`back!` was already used")
 
 accum!(x, Δ) = x .+ Δ
 accum!(x::AbstractArray, Δ) = (x .+= Δ)
 
+function back_(x::Tracked, Δ)
+    if isdefined(x, :grad)
+      x.grad = accum!(x.grad, Δ)
+    else
+      x.grad = Δ
+    end
+    return
+end
+
+back_(::Nothing, Δ) = return
+
 function back(x::Tracked, Δ, once)
-  x.isleaf && (x.grad = accum!(x.grad, Δ); return)
-  ref = x.ref -= 1
-  grad = if isdefined(x, :grad)
-    x.grad = accum!(x.grad, Δ)
-  elseif ref > 0
-    x.grad = Δ
-  else
-    Δ
-  end
-  if ref == 0
-    back_(x.f, grad, once)
-    once && !x.isleaf && (x.f = Call(missing, ()))
-  end
-  return
+    seen = Set{UInt64}(objectid(x))
+    if isdefined(x, :grad)
+      x.grad = zero_grad!(x.grad)
```

Comment:
Why are we calling this here? Seems like we should get the right behaviour without this as well.

Reply:
This is previously done in the `scan` https://github.com/FluxML/Tracker.jl/blob/b460dcf5199c7d64ec5ea9e2392fa131f7369b2e/src/back.jl#L19 but now `scan` is removed.